### PR TITLE
Use a dropdown (in preact) to select mutations from

### PR DESCRIPTION
**Pull request by anna-parker**
resolves https://github.com/GenSpectrum/dashboard-components/issues/482

### Summary

Instead of having to press the `+` button after typing in a mutation we parse the user's input and check if it is a valid mutation. If the input is valid we show that option in a dropdown. It can be selected from the dropdown, adding it to the list of selected mutation filters. This is more intuitive for users. 

Covspectrum uses the same approach. 

This is a preact reimplementation of `asyncSelect` from [react-select](https://react-select.com/home). 

### Screenshot

image

### PR Checklist

- [x] All necessary documentation has been adapted.
- [x] The implemented feature is covered by an appropriate test.
- [x] Potential follow up: allow selection of options in dropdown with the enter keyboard option.

### CODE REVIEW

**Review by fengelniederhammer — CHANGES_REQUESTED** [review]
As discussed: I think it would be simpler and equally good if we don't try to implement another `AsyncSelect`. I'd propose to have a dedicated mutation input field that doesn't aim to be general purpose, but is just a text input with a single "autocomplete option", if the value is a valid mutation.

IMO keyboard navigation is a must. There should be some way to hit "enter" and submit the muation. I'm not sure how a good solution would look like. `<datalist>` doesn't work, because it doesn't have events by itself. Instead the corresponding input element gets `input` event -> you can't differentiate whether an option was selected or whether something was typed in the input element.

(Most of my other comments are probably irrelevant now)

File: components/src/preact/mutationFilter/asyncselect.tsx [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        console.log('handleBlur started');
+        if (!wrapperRef.current?.contains(event.relatedTarget as Node)) {
+            console.log('handleBlur');
+            onBlur?.();
+            setMenuVisible(false);
+        }
+    };
+
+    return (
+        <div className={className} style={{ position: 'relative' }} ref={wrapperRef} tabIndex={-1}>
```
[the remark this answers is not in the document]
Comment (anna-parker) [reply components/src/preact/mutationFilter/asyncselect.tsx]: done :-)

File: components/src/preact/mutationFilter/asyncselect.tsx [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    // Load options asynchronously based on input value
+    useEffect(() => {
+        const loadAsyncSearchOptions = async () => {
+            setLoading(true);
+            const result = await loadOptions(inputValue);
+            setSearchOptions(result);
+            setLoading(false);
+        };
+
+        if (inputValue) loadAsyncSearchOptions();
```
[the remark this answers is not in the document]
Comment (anna-parker) [reply components/src/preact/mutationFilter/asyncselect.tsx]: done :-)

File: components/src/preact/mutationFilter/asyncselect.tsx
[the remark this answers is not in the document]
Comment (anna-parker) [reply components/src/preact/mutationFilter/asyncselect.tsx:1]: done :-)

**Comment by anna-parker** [comment]
@fengelniederhammer I now modified the async-select component to treat users clicking enter the same as clicking on an item. Should I leave the async-select component as is now and just add some more general tests for it? Or would you still prefer I try to make to his one component that is specifically for selecting mutations?

File: components/src/preact/mutationFilter/async-select.tsx [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+    return (
+        <div className={`relative ${className}`} ref={wrapperRef} tabIndex={-1}>
+            <input
+                type='text'
+                ref={inputRef}
+                className={className}
+                placeholder={placeholder}
+                value={inputValue}
+                onInput={(e: Event) => onInputChange((e.target as HTMLInputElement).value)}
```
Comment (anna-parker) [inline components/src/preact/mutationFilter/async-select.tsx:118]: replace with onKeyDown()

Comment (anna-parker) [reply components/src/preact/mutationFilter/async-select.tsx:118]: TODO: add a test that enter works as planned

**Comment by anna-parker** [comment]
I think it is easiest to close this PR as I added more features and tests in https://github.com/GenSpectrum/dashboard-components/pull/500 and it includes this functionality.

**Pull request closed without merging**